### PR TITLE
Add tidwall/buntdb instrumentation

### PR DESCRIPTION
**Pull request by TomRoSystems**
This new module instruments the `github.com/tidwall/buntdb` package.

- [x] Comply with OpenTelemetry HTTP semantic conventions
- [x] Package documentation
- [x] Exported objects are documented
- [x] Tests and linting are passing
- [x] Includes dependabot config
- [x] Library added to README.md
- [x] CHANGELOG.md updated

### CODE REVIEW

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/go.mod [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+go 1.15
+
+require (
+	github.com/signalfx/splunk-otel-go v0.6.0
+	github.com/tidwall/buntdb v1.2.7
+	go.opentelemetry.io/otel v1.2.0
+	go.opentelemetry.io/otel/trace v1.2.0
+)
+
+// replace github.com/signalfx/signalfx-go-tracing => ../../../
```
Comment (pellared) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/go.mod:12]: why is it commented?

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/go.mod [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/signalfx/signalfx-go-tracing/contrib/tidwall/buntdb/splunkbuntdb v0.0.0-00010101000000-000000000000
+	github.com/stretchr/testify v1.7.0
+	github.com/tidwall/buntdb v1.2.7
+	go.opentelemetry.io/otel v1.2.0
+	go.opentelemetry.io/otel/sdk v1.2.0
+	go.opentelemetry.io/otel/trace v1.2.0
+)
+
+// replace github.com/signalfx/signalfx-go-tracing/contrib/tidwall/buntdb/splunkbuntdb => /Users/trojek/github/splunk-otel-go/instrumentation/github.com/tidwall/buntdb/splunkbuntdb
```
Comment (pellared) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/go.mod:15]: I think you can remove this line 😉

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+	c.defaultStartOpts = []trace.SpanStartOption{
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(attrs...),
+	}
+
+	if c.tracer == nil {
+		c.tracer = otel.Tracer(
+			instrumentationName,
+			trace.WithInstrumentationVersion(splunkotel.Version()),
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:58]:
```suggestion
			trace.WithInstrumentationVersion(splunkotel.Version()),
			trace.WithSchemaURL(semconv.SchemaURL),
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+// resolveTracer returns an OTel tracer from the appropriate TracerProvider.
+//
+// If the passed context contains a span, the TracerProvider that created the
+// tracer that created that span will be used. Otherwise, the TracerProvider
+// from c is used.
+func (c *config) resolveTracer(ctx context.Context) trace.Tracer {
+	if span := trace.SpanFromContext(ctx); span.SpanContext().IsValid() {
+		return span.TracerProvider().Tracer(
+			instrumentationName,
+			trace.WithInstrumentationVersion(splunkotel.Version()),
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:91]:
```suggestion
			trace.WithInstrumentationVersion(splunkotel.Version()),
			trace.WithSchemaURL(semconv.SchemaURL),
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+		)
+	}
+	// There is a possibility that the config was not created with newConfig
+	// (i.e. new(Client)), try to handle this situation gracefully.
+	if c == nil || c.tracer == nil {
+		return otel.Tracer(
+			instrumentationName,
+			trace.WithInstrumentationVersion(splunkotel.Version()),
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:99]:
```suggestion
			trace.WithInstrumentationVersion(splunkotel.Version()),
			trace.WithSchemaURL(semconv.SchemaURL),
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+		copy(o[len(c.defaultStartOpts):], opts)
+	}
+
+	extraAttrs := []attribute.KeyValue{
+		semconv.DBOperationKey.String(spanName),
+	}
+
+	traceAttr := trace.WithAttributes(extraAttrs...)
+
+	o = append(o, traceAttr)
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:123]: Allocating the known capacity of the options slice from the start will ensure the underlying array is the exact size needed.


```suggestion
		o = make([]trace.SpanStartOption, len(opts), len(opts)+1)
		copy(o, opts)
	} else {
		n := len(c.defaultStartOpts)+len(opts)
		o = make([]trace.SpanStartOption, n, n+1)
		copy(o, c.defaultStartOpts)
		copy(o[len(c.defaultStartOpts):], opts)
	}
	o = append(o, trace.WithAttributes(semconv.DBOperationKey.String(spanName)))
```

It might also make more sense to pass `semconv.DBOperationKey.String(spanName))` as an argument to this method. This will avoid this sizing logic and ensure future use of `withSpan` does not mistakenly have non-DB operation spans annotated as DB operations.

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+	name := spanName
+	_, span := c.resolveTracer(c.ctx).Start(c.ctx, name, o...)
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:126]: Is `name` needed?

```suggestion
	_, span := c.resolveTracer(c.ctx).Start(c.ctx, spanName, o...)
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	o(c)
+}
+
+// WithTracerProvider returns an Option that sets the TracerProvider used with
+// this instrumentation library.
+func WithTracerProvider(tp trace.TracerProvider) Option {
+	return optionFunc(func(c *config) {
+		c.tracer = tp.Tracer(
+			instrumentationName,
+			trace.WithInstrumentationVersion(splunkotel.Version()),
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:155]:
```suggestion
			trace.WithInstrumentationVersion(splunkotel.Version()),
			trace.WithSchemaURL(semconv.SchemaURL),
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+func WithAttributes(attr []attribute.KeyValue) Option {
+	return optionFunc(func(c *config) {
+		c.defaultStartOpts = append(
+			c.defaultStartOpts,
+			trace.WithAttributes(attr...),
+		)
+	})
+}
+
+// WithContext sets the context for the transaction.
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:171]:
```suggestion
// WithContext sets the parent context for spans.
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+package splunkbuntdb_test
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+
+	buntdb "github.com/signalfx/signalfx-go-tracing/contrib/tidwall/buntdb/splunkbuntdb"
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go:24]: Please do not alias this. Leaving it as `splunkbuntdb` will clearly communicate in the example where the `splunkbuntdb` is actually used.

```suggestion
	"github.com/signalfx/signalfx-go-tracing/contrib/tidwall/buntdb/splunkbuntdb"
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+)
+
+// name is the Tracer name used to identify this instrumentation library.
+const name = "splunkdb"
+
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go:29]:
```suggestion
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+	err = db.View(func(tx *buntdb.Tx) error {
+		err := tx.Ascend("", func(key, value string) bool {
+			fmt.Printf("key: %s, value: %s\n", key, value)
+			return true
+		})
+		return err
+	})
+
+	span.End()
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go:51]: This does not look like compile-able code. Is this a leftover?

```suggestion
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/go.mod (hunk 1 above)
Comment (MrAlias) [reply instrumentation/github.com/tidwall/buntdb/splunkbuntdb/go.mod:12]: Agreed. Please uncomment and fix the relative path.

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"context"
+	//	"os"
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:19]:
```suggestion
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/tidwall/buntdb"
+
+	"go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/sdk/trace/tracetest"
+
+	global "go.opentelemetry.io/otel"
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:31]: Why alias this?

```suggestion
	"go.opentelemetry.io/otel"
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+	ctx, cancel = context.WithDeadline(ctx, d)
+	t.Cleanup(cancel)
+	return ctx
+}
+
+func testUpdate(t *testing.T, name string, f func(tx *splunkbuntdb.Tx) error) {
+	sr := tracetest.NewSpanRecorder()
+	tp := trace.NewTracerProvider(trace.WithSpanProcessor(sr))
+
+	global.SetTracerProvider(tp)
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:389]: Instead of setting a global `TracerProvider` which can have side effects, please instead explicitly pass the `TracerProvider` with the `WithTracerProvider` option?

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+	err := db.Update(f)
+	assert.NoError(t, err)
+
+	ctx := withTestingDeadline(context.Background(), t)
+	require.NoError(t, tp.Shutdown(ctx))
+	spans := sr.Ended()
+	require.Len(t, spans, 1)
+	span := spans[0]
+
+	assert.Len(t, spans, 1)
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:402]: This is asserted above.

```suggestion
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+	assert.Contains(t, span.Attributes(), semconv.DBSystemKey.String("buntdb"))
+	assert.Contains(t, span.Attributes(), semconv.DBOperationKey.String(name))
+	assert.Equal(t, span.Name(), name)
+}
+
+func testView(t *testing.T, name string, f func(tx *splunkbuntdb.Tx) error) {
+	sr := tracetest.NewSpanRecorder()
+	tp := trace.NewTracerProvider(trace.WithSpanProcessor(sr))
+
+	global.SetTracerProvider(tp)
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:413]: Similar here, please pass the `TracerProvider` explicitly.

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+	err := db.View(f)
+	assert.NoError(t, err)
+
+	ctx := withTestingDeadline(context.Background(), t)
+	require.NoError(t, tp.Shutdown(ctx))
+	spans := sr.Ended()
+	require.Len(t, spans, 1)
+	span := spans[0]
+
+	assert.Len(t, spans, 1)
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:426]:
```suggestion
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+	assert.Contains(t, span.Attributes(), semconv.DBSystemKey.String("buntdb"))
+	assert.Contains(t, span.Attributes(), semconv.DBOperationKey.String(name))
+	assert.Equal(t, span.Name(), name)
+}
+
+func getDatabase(t *testing.T, opts ...splunkbuntdb.Option) *splunkbuntdb.DB {
+	bdb, err := buntdb.Open(":memory:")
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:437]:
```suggestion
	require.NoError(t, err)
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 20]
```diff
@@ -0,0 +1,469 @@
...
+
+	err = bdb.CreateIndex("test-index", "regular:*", buntdb.IndexBinary)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:442]:
```suggestion
	require.NoError(t, bdb.CreateIndex("test-index", "regular:*", buntdb.IndexBinary))
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 21]
```diff
@@ -0,0 +1,469 @@
...
+
+	err = bdb.CreateSpatialIndex("test-spatial-index", "spatial:*", buntdb.IndexRect)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:447]:
```suggestion
	require.NoError(t, bdb.CreateSpatialIndex("test-spatial-index", "spatial:*", buntdb.IndexRect))
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+		tx.Set("spatial:b", "[2 2]", nil)
+		tx.Set("spatial:c", "[3 3]", nil)
+		tx.Set("spatial:d", "[4 4]", nil)
+		tx.Set("spatial:e", "[5 5]", nil)
+
+		return nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/buntdb_test.go:466]:
```suggestion
	require.NoError(t, bdb.Update(func(tx *buntdb.Tx) error {
		tx.Set("regular:a", "1", nil)
		tx.Set("regular:b", "2", nil)
		tx.Set("regular:c", "3", nil)
		tx.Set("regular:d", "4", nil)
		tx.Set("regular:e", "5", nil)

		tx.Set("spatial:a", "[1 1]", nil)
		tx.Set("spatial:b", "[2 2]", nil)
		tx.Set("spatial:c", "[3 3]", nil)
		tx.Set("spatial:d", "[4 4]", nil)
		tx.Set("spatial:e", "[5 5]", nil)

		return nil
	}))
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/go.mod [hunk 23]
```diff
@@ -0,0 +1,12 @@
+module github.com/signalfx/splunk-otel-go/instrumentation/github.com/tidwall/buntdb/splunkbuntdb
+
+go 1.15
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/go.mod:3]:
```suggestion
go 1.16
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/go.mod [hunk 24]
```diff
@@ -0,0 +1,16 @@
+module github.com/signalfx/splunk-otel-go/instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test
+
+go 1.16
+
+require (
+	github.com/davecgh/go-spew v1.1.1 // indirect
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/test/go.mod:6]: Is this a needed dependency?

**Comment by MrAlias** [comment]
Please add dependabot configuration for the added modules (`make gendependabot`) and be sure to fix all test failures.

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+const instrumentationName = "github.com/signalfx/splunk-otel-go/instrumentation/github.com/tidwall/buntdb/splunkbuntdb"
+
+type config struct {
+	ctx              context.Context
+	tracer           trace.Tracer
+	defaultStartOpts []trace.SpanStartOption
+}
+
+func newConfig(options ...Option) *config {
+	var c config
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:38]:
```suggestion
	c := config{
		defaultStartOpts: []trace.SpanStartOption{
			trace.WithSpanKind(trace.SpanKindClient),
			trace.WithAttributes(semconv.DBSystemKey.String("buntdb")),
		},
	}
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+
+	if c.ctx == nil {
+		c.ctx = context.Background()
+	}
+
+	return &c
+}
+
+// copyConfig is used internally to create new config instance
+func copyConfig(c *config) *config {
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply instrumentation/github.com/tidwall/buntdb/splunkbuntdb/config.go:70]: https://github.com/signalfx/splunk-otel-go/pull/226/files#diff-893f567a835c15c51905b5ea92074a2530f981c91faafc15afe230ecd3bc2290R67-R79

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/buntdb.go [hunk 27]
```diff
@@ -0,0 +1,325 @@
...
+	return tx.cfg.withSpan("Ascend", func() error {
+		return tx.Tx.Ascend(index, iterator)
+	})
+}
+
+// AscendEqual calls the underlying Tx.AscendEqual and traces the query.
+func (tx *Tx) AscendEqual(index, pivot string, iterator func(key, value string) bool) error {
+	return tx.cfg.withSpan("AscendEqual", func() error {
+		return tx.Tx.AscendEqual(index, pivot, iterator)
+	})
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/buntdb.go:123]: E.g.

```suggestion
	}, trace.WithAttribute(semconv.DBOperationKey.String("AscendEqual")))
```

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+
+	err = db.View(func(tx *buntdb.Tx) error {
+		err := tx.Ascend("", func(key, value string) bool {
+			fmt.Printf("key: %s, value: %s\n", key, value)
+			return true
+		})
+		return err
+	})
+
+	// span.End()
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go:50]:
```suggestion
```

File: instrumentation/github.com/syndtr/goleveldb/leveldb/splunkleveldb/go.mod [hunk 29]
```diff
@@ -1,6 +1,6 @@
 module github.com/signalfx/splunk-otel-go/instrumentation/github.com/syndtr/goleveldb/leveldb/splunkleveldb
 
-go 1.16
+go 1.15
```
Comment (MrAlias) [inline instrumentation/github.com/syndtr/goleveldb/leveldb/splunkleveldb/go.mod:3]: Please isolate changes to the new splunkbuntdb package. If we want to make this change lets do so in a separate PR.

File: instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	err = db.View(func(tx *splunkbuntdb.Tx) error {
+		errIn := tx.Ascend("", func(key, value string) bool {
+			fmt.Printf("key: %s, value: %s\n", key, value)
+			return true
+		})
+		return errIn
+	})
+
```
Comment (MrAlias) [inline instrumentation/github.com/tidwall/buntdb/splunkbuntdb/example_test.go:47]: nitpick

```suggestion
	if err != nil {
		panic(err)
	}

	err = db.View(func(tx *splunkbuntdb.Tx) error {
		errIn := tx.Ascend("", func(key, value string) bool {
			fmt.Printf("key: %s, value: %s\n", key, value)
			return true
		})
		return errIn
	})
```

**Review by MrAlias — COMMENTED** [review]
The splunkbuntdb package needs a README [similar to existing ones](https://github.com/signalfx/splunk-otel-go/blob/main/instrumentation/github.com/syndtr/goleveldb/leveldb/splunkleveldb/README.md).

**Review by MrAlias — COMMENTED** [review]
Can we add tests for the `Open` function and both `WithContext` methods?

**Review by MrAlias — APPROVED** [review]
